app/javascript/components/explore/ScatterTab.js

```diff
@@ -0,0 +1,161 @@
+  }
+
+  // identify any repeat graphs
+  const newContextMap = getNewContextMap(scatterParams, plotlyContextMap.current)
+
+  useEffect(() => {
+    plotlyContextMap.current = newContextMap
+  }, [])
+
+  return <div className="row">
+    { scatterParams.map((params, index) => {
+      const isTwoColRow = isTwoColumn && !(index === 0 && firstRowSingleCol)
+      const key = getKeyFromScatterParams(params)
+
+      return <div className={isTwoColRow ? 'col-md-6' : 'col-md-12'} key={newContextMap[key]}>
+        <ScatterPlot
+          {...{
+            studyAccession, plotPointsSelected, isCellSelecting, updateScatterColor
+          }}
+          {...params}
+          dimensions={getPlotDimensions({
+            isMultiRow,
+            isTwoColumn: isTwoColRow,
+            hasTitle: true
+          })}
+        />
+      </div>
+    })}
+    { scatterParams.length > MAX_PLOTS &&
+      <div className="panel">
+        <span>Due to browser limitations, only 8 plots can be shown at one time.  Deselect some spatial groups</span>
+      </div>
+    }
+  </div>
+}
+
+/** returns an array of params objects suitable for passing into ScatterPlot components
+ * (one for each plot).  Also returns layout variables
+ * This Handles 6 permutations: (spatial / noSpatial) X (no / single / multigene)
+ */
+export function getScatterParams(exploreInfo, exploreParams, isGene, isMultiGene) {
+  let isTwoColumn = true
+  let isMultiRow = false
+  let firstRowSingleCol = false
+  const isSpatial = exploreParams.spatialGroups?.length > 0
+
+  const scatterParams = []
+  if (isMultiGene && !exploreParams.consensus) {
+    if (isSpatial) {
+      // for multi-gene spatial, show the reference spatial cluster at the top, then
+      // show the expression in the spatial files
+      // we don't support multi-gene multi-spatial, so only the first spatial group is shown
+      firstRowSingleCol = true
+      isMultiRow = true
+      // first plot is reference cluster, so no genes
+      scatterParams.push({ ...exploreParams, cluster: exploreParams.spatialGroups[0], genes: [] })
+      exploreParams.genes.forEach(gene => {
+        scatterParams.push({ ...exploreParams, cluster: exploreParams.spatialGroups[0], genes: [gene] })
+      })
+    } else {
+      // for non-spatial multigene, show each gene, with the cluster as the second plot
+      isMultiRow = true
+      exploreParams.genes.forEach((gene, index) => {
+        scatterParams.push({ ...exploreParams, genes:  [exploreParams.genes[index]] })
+        if (index === 0) {
+          scatterParams.push({ ...exploreParams, genes: [] })
+        }
+      })
+    }
+  } else if (isGene) {
+    if (isSpatial) {
+      // for single-gene spatial, show the gene expression and reference cluster, then the expresion
+      // overlaid over the spatial plot, then the spatial-cluster plot
+      scatterParams.push({ ...exploreParams })
+      scatterParams.push({ ...exploreParams, genes: [] })
```

Comment:
I _think_ I understand why the multiple assignment is happening here, including the empty `genes`, but some more explanation would be helpful (either in person or as comments).

Reply:
Having it in comments would help!  The rationale isn't clear to me.

Reply:
Added some comments here and above to hopefully make it clearer.  `{...foo}` is essentially ES6 for `clone(foo)`. And since ExploreTab is passing around param objects and modified param objects across a whole bunch of separate components, I wanted to always use clones to make sure we were never accidentally modifying the original exploreParams object